Fixed loading unsafe structure from a static field.

MoFtZ opened the pull request:
Fixes https://github.com/m4rs-mt/ILGPU/issues/213.

When loading an unsafe structure from a static field, and the structure contains a fixed buffer, we need to load all the elements of the fixed buffer. This is accomplished by pinning the first element of the fixed buffer, then using unsafe code to copy the remaining elements one byte at a time.

Also had to alter equality comparison of `TypeNode` so that `PrimitiveType` and `PaddingType` are considered equal. Otherwise, the additional padding bytes would fail to consider a `PrimitiveValue` to be compatible with `PaddingType`.

MoFtZ commented on Src/ILGPU/IR/Construction/Structures.cs:
```diff
@@ -80,6 +84,34 @@ partial class IRBuilder
+
+                    // If the field value we just added was the start of a fixed buffer,
+                    // copy the remaining elements, byte by byte.
+                    if (field.GetCustomAttribute<FixedBufferAttribute>() != null)
```
QUESTION: Is this the best way to detected a fixed buffer?

m4rs-mt replied:
In general, there is no guarantee that all .Net compatible compilers attach this attribute. What we could to is to check whether the native `size` of a type enforces a custom "padding".

MoFtZ replied:
Changed to compare field size.

MoFtZ commented on Src/ILGPU/IR/Construction/Structures.cs:
```diff
@@ -80,6 +84,34 @@ partial class IRBuilder
+                            {
+                                byte* ptr = (byte*)handle.AddrOfPinnedObject();
+                                for (int j = initialBytes; j < numBytes; j++)
+                                {
```
QUESTION: Do I need to worry about endian-ness?

m4rs-mt replied:
I guess this should work 🤓

m4rs-mt commented on Src/ILGPU/IR/Types/TypeNode.cs:
```diff
@@ -301,5 +301,27 @@ public T As<T>(ILocation location)
+        /// <param name="left">The left type node.</param>
+        /// <param name="right">The right type node.</param>
+        /// <returns>True, if the left and right type nodes are the same.</returns>
+        public static bool operator ==(TypeNode left, TypeNode right) =>
```
I am not sure whether we should really implement these operators as the compiler is based on the invariant that if the types are the "same" (using `==`) they actually represent the same type. Is this really required?

MoFtZ replied:
I've changed the PR to handle this differently. It only affect the one method in `StructureValues.cs`, so I've just changed the assertion.

MoFtZ commented:
Replaced the check for `FixedBufferAttribute` with a check on field size, for padding.

Reverted changes to `TypeNode` equality and replaced with non-intrusive change to assertion.

Added new assertion for `PaddingType.BasicValueType == BasicValueType.Int8`.
